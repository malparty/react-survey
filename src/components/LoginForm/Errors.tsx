import { FormikProps } from 'formik';
import { LoginFormValues } from '.';
import BaseAlert from '../BaseAlert';

const LoginFormErrors = ({ errors, touched }: FormikProps<LoginFormValues>) => {
  const getErrorMessages = (): JSX.Element[] => {
    const errorMessages = Reflect.ownKeys(errors).map((key) => {
      return (
        Reflect.get(errors, key) &&
        Reflect.get(touched, key) && <div key={key.toString()}> {Reflect.get(errors, key)}</div>
      );
    });
    return errorMessages.filter((error) => error);
  };

  const errorMessages = getErrorMessages();

  return errorMessages.length > 0 ? <BaseAlert>{errorMessages}</BaseAlert> : null;
};

export default LoginFormErrors;
