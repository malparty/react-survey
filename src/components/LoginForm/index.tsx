import nimbleLogo from '@assets/images/nimble-logo.png';
import { Formik, Field, Form, FormikHelpers } from 'formik';
import * as Yup from 'yup';
import LoginFormErrors from './Errors';
import { useTranslation } from 'react-i18next';

export interface LoginFormValues {
  email: string;
  password: string;
}

export const submitLoginForm = async (values: LoginFormValues, { setSubmitting }: FormikHelpers<LoginFormValues>) => {
  await new Promise((resolve) => setTimeout(resolve, 2000)); // Will be replaced by API call in BackEnd Task

  console.log(JSON.stringify(values, null, 2)); // Will be replaced by authentication & redirection in BackEnd Task
  setSubmitting(false);
};

const LoginForm = () => {
  const { t } = useTranslation();

  const LoginSchema = Yup.object().shape({
    email: Yup.string().email(t('login.errors.email_invalid')).required(t('login.errors.email_required')),
    password: Yup.string()
      .min(8, t('login.errors.password_too_short'))
      .max(32, t('login.errors.password_too_long'))
      .required(t('login.errors.password_required')),
  });

  return (
    <div className="login-form">
      <img className="login-form__header-image" src={nimbleLogo} alt={t('login.nimble_logo_alt')} />
      <div className="login-form__header-title">{t('login.title')}</div>
      <Formik
        initialValues={{
          email: '',
          password: '',
        }}
        onSubmit={submitLoginForm}
        validationSchema={LoginSchema}
      >
        {(formData) => (
          <Form>
            <LoginFormErrors {...formData} />
            <div className="login-form__field">
              <label className="login-form__label" htmlFor="email">
                {t('login.email')}
              </label>
              <Field className="login-form__input form-control" type="email" name="email" id="email" />
            </div>
            <div className="login-form__field">
              <label className="login-form__label" htmlFor="password">
                {t('login.password')}
              </label>
              <Field className="login-form__input form-control" type="password" name="password" id="password" />
            </div>
            <div className="login-form__field">
              <button className="btn btn-light" type="submit">
                {t('login.submit')}
              </button>
            </div>
          </Form>
        )}
      </Formik>
    </div>
  );
};

export default LoginForm;
