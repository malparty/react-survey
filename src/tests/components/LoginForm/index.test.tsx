import { render } from '@testing-library/react';
import LoginForm, { LoginFormValues, submitLoginForm } from '@src/components/LoginForm';
import { FormikHelpers } from 'formik';

describe('LoginForm', () => {
  it('correctly renders', () => {
    const loginForm = render(<LoginForm />);

    expect(loginForm).toMatchSnapshot();
  });

  it('displays an email field with its label', () => {
    const loginForm = render(<LoginForm />);
    expect(loginForm.getByLabelText('login.email')).toBeVisible();
  });

  it('displays a password field with its label', () => {
    const loginForm = render(<LoginForm />);

    expect(loginForm.getByLabelText('login.password')).toBeVisible();
  });

  it('displays a Sign In button', () => {
    const loginForm = render(<LoginForm />);

    expect(loginForm.getByText('login.title')).toBeVisible();
  });

  it('calls the setSubmitting callback after submitLoginForn is called', async () => {
    const values = { email: 'test@email.com', password: '012345678' };

    const formikHelpers = { setSubmitting: jest.fn() } as unknown as FormikHelpers<LoginFormValues>;

    await submitLoginForm(values, formikHelpers);

    expect(formikHelpers.setSubmitting).toBeCalledWith(false);
  });
});
