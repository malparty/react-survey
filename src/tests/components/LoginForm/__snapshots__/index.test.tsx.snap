// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm correctly renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-form"
      >
        <img
          alt="login.nimble_logo_alt"
          class="login-form__header-image"
          src="nimble-logo.png"
        />
        <div
          class="login-form__header-title"
        >
          login.title
        </div>
        <form
          action="#"
        >
          <div
            class="login-form__field"
          >
            <label
              class="login-form__label"
              for="email"
            >
              login.email
            </label>
            <input
              class="login-form__input form-control"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
          <div
            class="login-form__field"
          >
            <label
              class="login-form__label"
              for="password"
            >
              login.password
            </label>
            <input
              class="login-form__input form-control"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="login-form__field"
          >
            <button
              class="btn btn-light"
              type="submit"
            >
              login.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-form"
    >
      <img
        alt="login.nimble_logo_alt"
        class="login-form__header-image"
        src="nimble-logo.png"
      />
      <div
        class="login-form__header-title"
      >
        login.title
      </div>
      <form
        action="#"
      >
        <div
          class="login-form__field"
        >
          <label
            class="login-form__label"
            for="email"
          >
            login.email
          </label>
          <input
            class="login-form__input form-control"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="login-form__field"
        >
          <label
            class="login-form__label"
            for="password"
          >
            login.password
          </label>
          <input
            class="login-form__input form-control"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="login-form__field"
        >
          <button
            class="btn btn-light"
            type="submit"
          >
            login.submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
